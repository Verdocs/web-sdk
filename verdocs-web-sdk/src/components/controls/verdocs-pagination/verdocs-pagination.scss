@import '../../../theme.scss';

verdocs-pagination {
  gap: 8px;
  display: flex;
  flex-direction: row;
  font-family: $primary-font;

  .page-button {
    height: 38px;
    display: flex;
    flex: 0 0 38px;
    cursor: pointer;
    font-size: 16px;
    font-weight: 600;
    border-radius: 4px;
    align-items: center;
    justify-content: center;
    color: $verdocs-dark-blue;
    border: 1px solid $verdocs-purple;

    svg {
      width: 18px;
      height: 18px;
    }

    &.selected {
      cursor: default;
      color: $verdocs-grey-4;
      background: $verdocs-purple;
    }
  }

  .ellipsis {
    width: 38px;
    height: 38px;
    display: flex;
    flex: 0 0 38px;
    line-height: 32px;
    justify-content: center;
    color: $verdocs-dark-blue;
  }
}
