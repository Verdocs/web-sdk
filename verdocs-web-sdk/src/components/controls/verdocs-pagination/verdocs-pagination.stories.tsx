import {html} from 'lit-html';
import {Meta} from '@storybook/web-components';

export default {
  title: 'Controls/Pagination',
  component: 'verdocs-pagination',
  args: {
    selectedPage: 0,
    itemCount: 37,
    perPage: 10,
    argTypes: {
      onSelectPage: {action: 'selectPage'},
    },
  },
} as Meta;

export const Pagination = ({selectedPage, itemCount, perPage, onSelectPage}) => html`<verdocs-pagination
  .selectedPage=${selectedPage}
  .itemCount=${itemCount}
  .perPage=${perPage}
  @selectPage=${onSelectPage}
/>`;
