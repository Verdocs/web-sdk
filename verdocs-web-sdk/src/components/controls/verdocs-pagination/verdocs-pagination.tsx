import {integerSequence} from '@verdocs/js-sdk';
import {Host, FunctionalComponent} from '@stencil/core';
import {Component, Prop, Element, h, Event, EventEmitter} from '@stencil/core';

const ChevronDoubleLeft = `<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor"><path stroke-linecap="round" stroke-linejoin="round" d="M18.75 19.5l-7.5-7.5 7.5-7.5m-6 15L5.25 12l7.5-7.5" /></svg>`;

const ChevronDoubleRight = `<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor"><path stroke-linecap="round" stroke-linejoin="round" d="M11.25 4.5l7.5 7.5-7.5 7.5m-6-15l7.5 7.5-7.5 7.5" /></svg>`;

type TPageNumber = number | 'first' | 'last';

const VISIBLE_PAGES = 5;

const PageButton: FunctionalComponent<{page: TPageNumber; selected: number; onClick: (page: TPageNumber) => void}> = ({page, selected, onClick}) => {
  let label;
  if (page === 'first') {
    label = ChevronDoubleLeft;
  } else if (page === 'last') {
    label = ChevronDoubleRight;
  } else {
    label = String(page + 1);
  }

  return <div class={`page-button ${page === selected ? 'selected' : ''}`} onClick={() => onClick(page)} innerHTML={label} />;
};

@Component({
  tag: 'verdocs-pagination',
  styleUrl: 'verdocs-pagination.scss',
})
export class VerdocsQuickFilter {
  @Element()
  el: HTMLElement;

  /**
   * The currently selected page.
   */
  @Prop({mutable: true, reflect: true}) selectedPage: number = 0;

  /**
   * The total number of items.
   */
  @Prop() itemCount: number = 1;

  /**
   * The number of displayed per page.
   */
  @Prop() perPage: number = 10;

  /**
   * Event fired when the selected page changes. The new page number is included in the event.
   */
  @Event({composed: true}) selectPage: EventEmitter<{selectedPage: number}>;

  handleSelectPage(page: number) {
    this.selectedPage = page;
    this.selectPage?.emit({selectedPage: page});
  }

  render() {
    const firstPage = Math.max(0, this.selectedPage - 2);
    const numPages = this.itemCount > 0 ? Math.ceil(this.itemCount / this.perPage) : 0;
    const pagesToDisplay = integerSequence(0, Math.ceil(this.itemCount / this.perPage)).slice(firstPage, firstPage + VISIBLE_PAGES);

    return (
      <Host>
        {this.selectedPage > 0 && <PageButton page={'first'} selected={this.selectedPage} onClick={() => this.handleSelectPage(0)} />}

        {firstPage > 0 && <div class="ellipsis">...</div>}

        {pagesToDisplay.map(pageNumber => (
          <PageButton page={pageNumber} selected={this.selectedPage} onClick={page => this.handleSelectPage(+page)} />
        ))}

        {this.selectedPage < numPages - 1 && <div class="ellipsis">...</div>}

        {this.selectedPage < numPages - 1 && <PageButton page={'last'} selected={this.selectedPage} onClick={() => this.handleSelectPage(numPages - 1)} />}
      </Host>
    );
  }
}
