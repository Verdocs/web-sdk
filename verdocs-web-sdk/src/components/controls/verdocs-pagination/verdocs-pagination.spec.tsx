import { newSpecPage } from '@stencil/core/testing';
import { h } from '@stencil/core';
import { VerdocsQuickFilter } from './verdocs-pagination';

describe('verdocs-pagination', () => {
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [VerdocsQuickFilter],
      html: `<verdocs-pagination></verdocs-pagination>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with custom selectedPage, itemCount, and perPage', async () => {
    const page = await newSpecPage({
      components: [VerdocsQuickFilter],
      template: () => <verdocs-pagination selectedPage={2} itemCount={100} perPage={20}></verdocs-pagination>,
    });
    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.selectedPage).toBe(2);
    expect(page.rootInstance.itemCount).toBe(100);
    expect(page.rootInstance.perPage).toBe(20);
  });

  it('emits selectPage event on page button click', async () => {
    const page = await newSpecPage({
      components: [VerdocsQuickFilter],
      template: () => <verdocs-pagination selectedPage={1} itemCount={50} perPage={10}></verdocs-pagination>,
    });
    const spy = jest.fn();
    page.root.addEventListener('selectPage', spy);

    // Simulate clicking the second page button
    const pageButtons = page.root.querySelectorAll('.page-button');
    if (pageButtons.length > 1) {
      pageButtons[1].dispatchEvent(new Event('click', { bubbles: true }));
      expect(spy).toHaveBeenCalled();
    }
  });
});
