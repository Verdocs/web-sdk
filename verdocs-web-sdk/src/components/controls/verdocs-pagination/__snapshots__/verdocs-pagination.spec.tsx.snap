// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-pagination renders with custom selectedPage, itemCount, and perPage 1`] = `
<verdocs-pagination selected-page="2">
  <div class="page-button">
    <svg fill="none" stroke="currentColor" stroke-width="1.5" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
      <path d="M18.75 19.5l-7.5-7.5 7.5-7.5m-6 15L5.25 12l7.5-7.5" stroke-linecap="round" stroke-linejoin="round"></path>
    </svg>
  </div>
  <div class="page-button">
    1
  </div>
  <div class="page-button">
    2
  </div>
  <div class="page-button selected">
    3
  </div>
  <div class="page-button">
    4
  </div>
  <div class="page-button">
    5
  </div>
  <div class="ellipsis">
    ...
  </div>
  <div class="page-button">
    <svg fill="none" stroke="currentColor" stroke-width="1.5" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
      <path d="M11.25 4.5l7.5 7.5-7.5 7.5m-6-15l7.5 7.5-7.5 7.5" stroke-linecap="round" stroke-linejoin="round"></path>
    </svg>
  </div>
</verdocs-pagination>
`;

exports[`verdocs-pagination renders with default props 1`] = `
<verdocs-pagination selected-page="0">
  <div class="page-button selected">
    1
  </div>
  <div class="page-button">
    2
  </div>
</verdocs-pagination>
`;
